File: app/src/main/java/com/allybros/superego/activity/EditProfileActivity.java
package com.allybros.superego.activity;

import android.content.Context;
import android.content.Intent;
import android.graphics.Bitmap;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.net.Uri;
import android.os.Bundle;
import android.provider.MediaStore;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.Log;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageView;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.constraintlayout.widget.ConstraintLayout;

import com.allybros.superego.R;
import com.allybros.superego.api.ChangeInfoTask;
import com.allybros.superego.api.ImageChangeTask;
import com.allybros.superego.api.response.ApiStatusResponse;
import com.allybros.superego.unit.ConstantValues;
import com.allybros.superego.util.ClientContextUtil;
import com.allybros.superego.util.InputMethodWatcher;
import com.allybros.superego.util.SessionManager;
import com.allybros.superego.widget.SegoEditText;
import com.google.android.material.snackbar.BaseTransientBottomBar;
import com.google.android.material.snackbar.Snackbar;
import com.squareup.picasso.Picasso;

import java.io.IOException;

import de.hdodenhof.circleimageview.CircleImageView;
import me.zhanghai.android.materialprogressbar.MaterialProgressBar;

import static com.allybros.superego.util.HelperMethods.imageToString;

public class EditProfileActivity extends AppCompatActivity {
    private MaterialProgressBar progressEditProfile;
    private SegoEditText etUsername;
    private SegoEditText etEmail;
    private EditText etBio;
    private ImageView ivChangeAvatarIcon;
    private CircleImageView ivAvatar;
    private ConstraintLayout editProfileLayout;
    private Button btnSaveProfile;
    private static final int IMG_REQUEST = 1; //Needs for image selection from local storage

    private InputMethodWatcher inputMethodWatcher;

    @Override
    protected void onCreate(final Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_edit_profile);

        initializeComponents();
        setupUi();
        setupTextWatchers();
        initInputMethodWatcher();
    }

    private void initializeComponents() {
        editProfileLayout = findViewById(R.id.editProfileLayout);
        progressEditProfile = findViewById(R.id.progressEditProfile);
        findViewById(R.id.cardFormEditProfile);
        ivChangeAvatarIcon = findViewById(R.id.ivChangeAvatar);
        findViewById(R.id.ivBack);
        etUsername = findViewById(R.id.etUsername);
        etEmail = findViewById(R.id.etEmail);
        etBio = findViewById(R.id.etInformation);
        ivAvatar = findViewById(R.id.ivUserAvatarEditProfile);
        btnSaveProfile = findViewById(R.id.btnSaveProfile);
    }

    private void setupUi() {

        //Set view components
        etEmail.setText(SessionManager.getInstance().getUser().getEmail());
        etUsername.setText(SessionManager.getInstance().getUser().getUsername());
        etBio.setText(SessionManager.getInstance().getUser().getUserBio());


        // Check internet connection
        ConnectivityManager cm = (ConnectivityManager) getApplicationContext().getSystemService(Context.CONNECTIVITY_SERVICE);
        NetworkInfo activeNetwork = cm.getActiveNetworkInfo();
        boolean isConnected = activeNetwork != null && activeNetwork.isConnectedOrConnecting();
        //Load image
        String URL = SessionManager.getInstance().getUser().getImage();
        Picasso.get().load(URL).error(R.drawable.default_avatar).into(ivAvatar);

        if (!isConnected)
            Snackbar.make(editProfileLayout, R.string.error_no_connection, BaseTransientBottomBar.LENGTH_LONG).show();

        ivChangeAvatarIcon.setOnClickListener(this::onChangeAvatarClick);
        ivAvatar.setOnClickListener(this::onChangeAvatarClick);
        btnSaveProfile.setOnClickListener(view -> {
            if (inputMethodWatcher.isKeyboardShown()) {
                Log.d("Page changed", "Hide soft keyboard");
                inputMethodWatcher.hideSoftKeyboard();
            }

            // Check internet connection
            if (ClientContextUtil.isNetworkConnected(getApplicationContext())) {
                saveProfile();
            } else {
                Snackbar.make(editProfileLayout, R.string.error_no_connection, BaseTransientBottomBar.LENGTH_LONG).show();
            }
        });
    }

    public void onChangeAvatarClick(View view) {
        if (ClientContextUtil.isNetworkConnected(this)) {
            Intent intent = new Intent();
            intent.setType("image/*");
            intent.setAction(Intent.ACTION_GET_CONTENT);
            startActivityForResult(intent, IMG_REQUEST);
        } else {
            Snackbar.make(editProfileLayout, R.string.error_no_connection, BaseTransientBottomBar.LENGTH_LONG).show();
        }
    }

    private void setupTextWatchers() {
        etUsername.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
                //this method is empty
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                //this method is empty
            }

            @Override
            public void afterTextChanged(Editable s) {
                if (s.toString().isEmpty()) {
                    etUsername.setError(getString(R.string.error_username_empty));
                } else {
                    etUsername.clearError();
                }
            }
        });

        etEmail.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
                //this method is empty
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                //this method is empty
            }

            @Override
            public void afterTextChanged(Editable s) {
                if (s.toString().isEmpty()) {
                    etEmail.setError(getString(R.string.error_email_empty));
                } else {
                    etEmail.clearError();
                }
            }
        });

    }

    //Provides that cacth the results that come back from selectImage() function
    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        if (requestCode == IMG_REQUEST && resultCode == RESULT_OK && data != null) {
            Uri imagePath = data.getData();
            try {
                Bitmap bitmap = MediaStore.Images.Media.getBitmap(getContentResolver(), imagePath);
                int fileSize = bitmap.getByteCount();
                Log.d("SIZE:  ", "" + bitmap.getByteCount());
                if (fileSize < ConstantValues.MAX_FILE_SIZE) {
                    SessionManager.getInstance().getUser().setAvatar(bitmap);
                    ivAvatar.setImageBitmap(SessionManager.getInstance().getUser().getAvatar());
                    ivAvatar.setVisibility(View.INVISIBLE);
                    setProgressVisibility(true);
                    ImageChangeTask imageChangeTask = new ImageChangeTask(SessionManager.getInstance().getSessionToken(), imageToString(SessionManager.getInstance().getUser().getAvatar()));
                    imageChangeTask.setOnResponseListener(this::handleImageChangeTaskResponse);
                    imageChangeTask.execute(this);
                } else {
                    Snackbar.make(editProfileLayout, getString(R.string.error_invalid_file_size), BaseTransientBottomBar.LENGTH_LONG).show();
                }

            } catch (IOException e) {
                e.printStackTrace();
            }
        }
    }

    private void handleImageChangeTaskResponse(ApiStatusResponse response) {
        setProgressVisibility(false);
        ivAvatar.setVisibility(View.VISIBLE);
        Snackbar.make(editProfileLayout, response.getMessage(), BaseTransientBottomBar.LENGTH_LONG).show();
    }

    /**
     * Shows material progress bar and disables form.
     *
     * @param visible set true when progress view needs to be shown.
     */
    private void setProgressVisibility(boolean visible) {
        if (visible) {
            progressEditProfile.setVisibility(View.VISIBLE);
            etUsername.setEnabled(false);
            etEmail.setEnabled(false);
            etBio.setEnabled(false);
            btnSaveProfile.setEnabled(false);
        } else {
            progressEditProfile.setVisibility(View.INVISIBLE);
            etUsername.setEnabled(true);
            etEmail.setEnabled(true);
            etBio.setEnabled(true);
            btnSaveProfile.setEnabled(true);
        }
    }

    /**
     * Validate user information and send request to the API
     */
    private void saveProfile() {
        etUsername.clearError();
        etEmail.clearError();

        if (etUsername.getText().isEmpty()) {
            etUsername.setError(getString(R.string.error_username_empty));
        }
        if (etEmail.getText().isEmpty()) {
            etEmail.setError(getString(R.string.error_email_empty));
        }

        if (!etUsername.getText().isEmpty() && !etEmail.getText().isEmpty()) {
            setProgressVisibility(true);
            ChangeInfoTask changeInfoTask = new ChangeInfoTask(
                    SessionManager.getInstance().getSessionToken(),
                    etUsername.getText(),
                    etEmail.getText(),
                    etBio.getText().toString());
            changeInfoTask.setOnResponseListener(response -> handleChangeInfoTaskResponse(response,
                    etUsername.getText(),
                    etEmail.getText(),
                    etBio.getText().toString()));
            changeInfoTask.execute(this);
        }
    }

    private void handleChangeInfoTaskResponse(ApiStatusResponse response, String uid, String email, String bio) {
        setProgressVisibility(false);
        //Check status
        String text = response.getMessage();
        Snackbar.make(editProfileLayout, text, BaseTransientBottomBar.LENGTH_LONG).show();
        SessionManager.getInstance().updateLocalVariables(uid, email, bio);
    }

    public void onBackButtonPressed(View view) {
        getOnBackPressedDispatcher().onBackPressed();
    }

    /**
     * Initializes input method watcher for detecting virtual keyboard.
     */
    private void initInputMethodWatcher() {
        View contentRoot = ((ViewGroup) findViewById(R.id.editProfileLayout)).getChildAt(0);
        inputMethodWatcher = new InputMethodWatcher(contentRoot);
    }
}

File: app/src/main/java/com/allybros/superego/activity/editprofile/EditProfileActivity.kt
package com.allybros.superego.activity.editprofile

import android.widget.EditText
import com.allybros.superego.R
import android.content.Intent
import com.allybros.superego.unit.ErrorCodes
import com.google.android.material.snackbar.Snackbar
import com.allybros.superego.unit.ConstantValues
import android.net.ConnectivityManager
import com.squareup.picasso.Picasso
import com.google.android.material.snackbar.BaseTransientBottomBar
import android.text.TextWatcher
import android.text.Editable
import android.net.Uri
import android.os.Bundle
import android.os.PersistableBundle
import android.provider.MediaStore
import android.util.Log
import android.view.View
import androidx.activity.viewModels
import androidx.core.content.ContextCompat
import com.allybros.superego.base.SegoActivity
import com.allybros.superego.util.HelperMethods
import com.allybros.superego.util.SessionManager
import java.io.IOException
import com.allybros.superego.databinding.ActivityEditProfileBinding
import com.allybros.superego.unit.EditProfileRequest
import com.allybros.superego.unit.UploadImageRequest

class EditProfileActivity : SegoActivity<EditProfileVM, ActivityEditProfileBinding>() {
    private val IMG_REQUEST = 1 //Needs for image selection from local storage
    private var newImagePath: Uri? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        observeEvents()
        setupUi()
        setListeners()
        setupTextWatchers()
    }

    override fun getSegoViewModel(): EditProfileVM {
        val viewModel by viewModels<EditProfileVM>()
        return viewModel
    }

    override fun getLayout(): Int = R.layout.activity_edit_profile

    private fun observeEvents(){
        getViewModel().editProfileResponseLiveData.observe(this, {
            setProgressVisibility(false)
            when (it?.status) {
                ErrorCodes.SESSION_EXPIRED -> Snackbar.make(
                    binding.editProfileLayout,
                    applicationContext.getString(R.string.error_session_expired),
                    Snackbar.LENGTH_LONG
                ).show()
                ErrorCodes.USERNAME_NOT_LEGAL -> Snackbar.make(
                    binding.editProfileLayout,
                    applicationContext.getString(R.string.error_username_not_legal),
                    Snackbar.LENGTH_LONG
                ).show()
                ErrorCodes.USERNAME_ALREADY_EXIST -> Snackbar.make(
                    binding.editProfileLayout,
                    applicationContext.getString(R.string.error_username_taken),
                    Snackbar.LENGTH_LONG
                ).show()
                ErrorCodes.EMAIL_NOT_LEGAL -> Snackbar.make(
                    binding.editProfileLayout,
                    applicationContext.getString(R.string.error_email_not_legal),
                    Snackbar.LENGTH_LONG
                ).show()
                ErrorCodes.EMAIL_ALREADY_EXIST -> Snackbar.make(
                    binding.editProfileLayout,
                    applicationContext.getString(R.string.error_email_already_exist),
                    Snackbar.LENGTH_LONG
                ).show()
                ErrorCodes.SUCCESS -> Snackbar.make(
                    binding.editProfileLayout,
                    applicationContext.getString(R.string.message_process_succeed),
                    Snackbar.LENGTH_LONG
                ).show()
                ErrorCodes.SYSFAIL -> Snackbar.make(
                    binding.editProfileLayout,
                    applicationContext.getString(R.string.error_no_connection),
                    Snackbar.LENGTH_LONG
                ).show()
            }
        })

        getViewModel().editProfileResponseLiveData.observe(this, {
            setProgressVisibility(false)
            binding.ivUserAvatarEditProfile.visibility = View.VISIBLE
            when (it.status) {
                ErrorCodes.SUCCESS -> Snackbar.make(
                    binding.editProfileLayout,
                    applicationContext.getString(R.string.message_process_succeed),
                    Snackbar.LENGTH_LONG
                ).show()
                ErrorCodes.SYSFAIL -> Snackbar.make(
                    binding.editProfileLayout,
                    applicationContext.getString(R.string.error_no_connection),
                    Snackbar.LENGTH_LONG
                ).show()
                ErrorCodes.INVALID_FILE_EXTENSION -> Snackbar.make(
                    binding.editProfileLayout,
                    applicationContext.getString(R.string.error_invalid_file_type),
                    Snackbar.LENGTH_LONG
                ).show()
                ErrorCodes.INVALID_FILE_TYPE -> Snackbar.make(
                    binding.editProfileLayout,
                    applicationContext.getString(R.string.error_invalid_file_type),
                    Snackbar.LENGTH_LONG
                ).show()
                ErrorCodes.INVALID_FILE_SIZE -> Snackbar.make(
                    binding.editProfileLayout,
                    applicationContext.getString(R.string.error_invalid_file_size),
                    Snackbar.LENGTH_LONG
                ).show()
                ErrorCodes.FILE_WRITE_ERROR -> Snackbar.make(
                    binding.editProfileLayout,
                    applicationContext.getString(R.string.error_no_connection),
                    Snackbar.LENGTH_LONG
                ).show()
            }
        })
    }

    private fun setupUi() {

        //Set view components
        binding.etEmail.setText(SessionManager.getInstance().user.email)
        binding.etUsername.setText(SessionManager.getInstance().user.username)
        binding.etBio.setText(SessionManager.getInstance().user.userBio)


        // Check internet connection
        val cm = applicationContext.getSystemService(CONNECTIVITY_SERVICE) as ConnectivityManager
        val activeNetwork = cm.activeNetworkInfo
        val isConnected = activeNetwork != null && activeNetwork.isConnectedOrConnecting
        //Load image
        val URL = SessionManager.getInstance().user.image
        Picasso.get().load(URL).into(binding.ivUserAvatarEditProfile)
        if (!isConnected) Snackbar.make(
            binding.editProfileLayout,
            R.string.error_no_connection,
            BaseTransientBottomBar.LENGTH_LONG
        ).show()
    }

    private fun setListeners(){
        binding.ivChangeAvatar.setOnClickListener { // Check internet connection
            val cm = applicationContext.getSystemService(CONNECTIVITY_SERVICE) as ConnectivityManager
            val activeNetwork = cm.activeNetworkInfo
            val isConnected = activeNetwork != null && activeNetwork.isConnectedOrConnecting
            if (isConnected) {
                selectImage()
            } else {
                Log.d("CONNECTION", isConnected.toString())
                Snackbar.make(
                    binding.editProfileLayout,
                    R.string.error_no_connection,
                    BaseTransientBottomBar.LENGTH_LONG
                ).show()
            }
        }


        binding.btnSaveProfile.setOnClickListener { // Check internet connection
            val cm = applicationContext.getSystemService(CONNECTIVITY_SERVICE) as ConnectivityManager
            val activeNetwork = cm.activeNetworkInfo
            val isConnected = activeNetwork != null && activeNetwork.isConnectedOrConnecting
            if (isConnected) {
                saveProfile()
            } else {
                Log.d("CONNECTION", isConnected.toString())
                Snackbar.make(
                    binding.editProfileLayout,
                    R.string.error_no_connection,
                    BaseTransientBottomBar.LENGTH_LONG
                ).show()
            }
        }

        binding.ivBack.setOnClickListener { onBackPressed() }
        binding.tvOptionsTitle.setOnClickListener { onBackPressed() }
    }
    private fun setupTextWatchers() {
        binding.etUsername.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {
                //this method is empty
            }

            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                //this method is empty
            }

            override fun afterTextChanged(s: Editable) {
                if (s.toString().isEmpty()) {
                    setError(binding.etUsername, getString(R.string.error_username_empty))
                } else {
                    clearError(binding.etUsername)
                }
            }
        })
        binding.etEmail.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {
                //this method is empty
            }

            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                //this method is empty
            }

            override fun afterTextChanged(s: Editable) {
                if (s.toString().isEmpty()) {
                    setError(binding.etEmail, getString(R.string.error_email_empty))
                } else {
                    clearError(binding.etEmail)
                }
            }
        })
    }

    //Opens intent that provide selecting image from local storage
    private fun selectImage() {
        val intent = Intent()
        intent.type = "image/*"
        intent.action = Intent.ACTION_GET_CONTENT
        startActivityForResult(intent, IMG_REQUEST)
    }

    //Provides that cacth the results that come back from selectImage() function
    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == IMG_REQUEST && resultCode == RESULT_OK && data != null) {
            newImagePath = data.data
            try {
                val bitmap = MediaStore.Images.Media.getBitmap(contentResolver, newImagePath)
                val fileSize = bitmap.byteCount
                Log.d("SIZE:  ", "" + bitmap.byteCount)
                if (fileSize < ConstantValues.MAX_FILE_SIZE) {
                    SessionManager.getInstance().user.avatar = bitmap
                    binding.ivUserAvatarEditProfile.setImageBitmap(SessionManager.getInstance().user.avatar)
                    binding.ivUserAvatarEditProfile.visibility = View.INVISIBLE
                    setProgressVisibility(true)
                    getViewModel().uploadImage(UploadImageRequest(
                        SessionManager.getInstance().sessionToken,
                        HelperMethods.imageToString(SessionManager.getInstance().user.avatar)
                    ))
                } else {
                    Snackbar.make(
                        binding.editProfileLayout,
                        "Seçilen dosya boyutu çok büyük",
                        Snackbar.LENGTH_LONG
                    ).show()
                }
            } catch (e: IOException) {
                e.printStackTrace()
            }
        }
    }

    /**
     * Shows material progress bar and disables form.
     * @param visible set true when progress view needs to be shown.
     */
    private fun setProgressVisibility(visible: Boolean) {
        if (visible) {
            binding.progressEditProfile.visibility = View.VISIBLE
            binding.cardFormEditProfile.alpha = 0.5f
            binding.btnSaveProfile.isEnabled = false
        } else {
            binding.progressEditProfile.visibility = View.INVISIBLE
            binding.cardFormEditProfile.alpha = 1f
            binding.btnSaveProfile.isEnabled = true
        }
    }

    /**
     * Validate user information and send request to the API
     */
    private fun saveProfile() {
        clearError(binding.etUsername)
        clearError(binding.etEmail)
        if (binding.etUsername.text.toString().isEmpty()) {
            setError(binding.etUsername, getString(R.string.error_username_empty))
        }
        if (binding.etEmail.text.toString().isEmpty()) {
            setError(binding.etEmail, getString(R.string.error_email_empty))
        }
        if (binding.etUsername.text.toString().isNotEmpty() && binding.etEmail.text.toString().isNotEmpty()) {
            setProgressVisibility(true)

            getViewModel().editProfile(EditProfileRequest(
                sessionToken =SessionManager.getInstance().sessionToken,
                newUsername = binding.etUsername.text.toString(),
                newEmail = binding.etEmail.text.toString(),
                newUserBio = binding.etBio.text.toString()
            ))
        }
    }

    private fun setError(editText: EditText?, errorMessage: String) {
        editText!!.hint = errorMessage
        editText.background = ContextCompat.getDrawable(
            applicationContext,
            R.drawable.et_error_background
        )
    }

    private fun clearError(editText: EditText?) {
        editText!!.background = ContextCompat.getDrawable(
            applicationContext,
            R.drawable.et_background
        )
    }
}